### ISSUE TITLE
Exception: Detected a page unload event; asynchronous script execution does not work across page loads.

### ISSUE THREAD

**Issue by debanshuk** [angular/protractor#3756]
Bug report
---
On navigating to a new page from the client action (a button click),  an exception is raised when calling any `browser` API which calls `waitForAngular()`, on PhantomJS. It work fine (without any exception) on Chrome though. Here is the exception message:

**Detected a page unload event; asynchronous script execution does not work across page loads.**

Similar issue have been reported here in past too and all of them are in closed state. Following are the past issues which I found:

https://github.com/angular/protractor/issues/85 (As stated in the comments of this issue the bug has been resolved in 2014 in commit https://github.com/angular/protractor/commit/a0bd84b9a28ec92eccd2784f8b849388985a4480. Though it looks like it has some how reappeared.)
https://github.com/angular/protractor/issues/1651
https://github.com/angular/protractor/issues/560
https://github.com/angular/protractor/issues/1901

Following are various version details, test and config code and error details.

- Node Version: `4.2.6`
- Protractor Version: `4.0.11`
- Angular Version: `1.4.8`
- Browser(s): `PhantomJS 2.1.1, Chrome 54.0.2840.100 (64-bit)` [bug is only occurring in PhantomJS]
- Operating System `Arch Linux (64 bit) (linux 4.7.6-1)`

- Following is my protractor configuration file:
```javascript
exports.config = {
    framework: 'jasmine',
    seleniumAddress: 'http://localhost:4444/wd/hub',
    allScriptsTimeout: 200000,
    getPageTimeout: 200000,
    jasmineNodeOpts: {
        defaultTimeoutInterval: 300000
    },
    specs: [
        'specs/**/*.js'
    ],
    multiCapabilities: [
        {
            'browserName': 'phantomjs',

            /*
             * Can be used to specify the phantomjs binary path.
             * This can generally be ommitted if you installed phantomjs globally.
             */
            'phantomjs.binary.path': require('phantomjs-prebuilt').path,

            /*
             * Command line args to pass to ghostdriver, phantomjs's browser driver.
             * See https://github.com/detro/ghostdriver#faq
             */
            'phantomjs.ghostdriver.cli.args': ['--loglevel=DEBUG']
        },
        {
            'browserName': 'chrome'
        }
    ]
};
```

- Following the the test cases in which the error is occurring.
```javascript
var format = require('string-format');

describe('Login page', function() {
    it('should allow login using a user', function() {
        // Go to login page.
        browser.get(format('{baseUrl}/login', {baseUrl: browser.params.baseUrl}));

        // Populate username and password.
        [['user.username', 'xxxxxxx'], ['user.password', 'xxxxxxx']].forEach(function(value) {
            var input = element(by.model(value[0]));
            input.isDisplayed();
            input.sendKeys(value[1]);
        });

        // Click the sign-in button.
        element(by.xpath('//button/*[.="Sign In to App"]/..')).click();

        // Check URL of the page to make sure we are on dashboard.
        expect(browser.getCurrentUrl()).toEqual(format('{baseUrl}/dashboard', {baseUrl: format(browser.params.baseUrlWithDomain, {subDomain: 'xxxx'})}));
    });
});
```

-  Following is the error which occurred, during execution of last `browser.getCurrentUrl()` function (I know this because it works if I remove the function call):
```json
{
    "errorMessage": "Detected a page unload event; asynchronous script execution does not work across page loads.",
    "request":
    {
        "headers":
        {
            "Accept-Encoding": "gzip,deflate",
            "Connection": "Keep-Alive",
            "Content-Length": "1196",
            "Content-Type": "application/json; charset=utf-8",
            "Host": "localhost:12512",
            "User-Agent": "Apache-HttpClient/4.5.1 (Java/1.8.0_102)"
        },
        "httpVersion": "1.1",
        "method": "POST",
        "post": "{\"script\":\"try { return (function (rootSelector, ng12Hybrid, callback) {\\n  var el = document.querySelector(rootSelector);\\n\\n  try {\\n    if (!ng12Hybrid && window.getAngularTestability) {\\n      window.getAngularTestability(el).whenStable(callback);\\n      return;\\n    }\\n    if (!window.angular) {\\n      throw new Error('window.angular is undefined.  This could be either ' +\\n          'because this is a non-angular page or because your test involves ' +\\n          'client-side navigation, which can interfere with Protractor\\\\'s ' +\\n          'bootstrapping.  See http://git.io/v4gXM for details');\\n    }\\n    if (angular.getTestability) {\\n      angular.getTestability(el).whenStable(callback);\\n    } else {\\n      if (!angular.element(el).injector()) {\\n        throw new Error('root element (' + rootSelector + ') has no injector.' +\\n           ' this may mean it is not inside ng-app.');\\n      }\\n      angular.element(el).injector().get('$browser').\\n          notifyWhenNoOutstandingRequests(callback);\\n    }\\n  } catch (err) {\\n    callback(err.message);\\n  }\\n}).apply(this, arguments); }\\ncatch(e) { throw (e instanceof Error) ? e : new Error(e); }\",\"args\":[\"body\",false]}",
        "url": "/execute_async",
        "urlParsed":
        {
            "anchor": "",
            "query": "",
            "file": "execute_async",
            "directory": "/",
            "path": "/execute_async",
            "relative": "/execute_async",
            "port": "",
            "host": "",
            "password": "",
            "user": "",
            "userInfo": "",
            "authority": "",
            "protocol": "",
            "source": "/execute_async",
            "queryKey": {},
            "chunks":
            [
                "execute_async"
            ]
        },
        "urlOriginal": "/session/252fa430-aebd-11e6-9e45-657168d95ebc/execute_async"
    }
}
```

-  This error only occurs when running the above test with PhantomJS. The test runs fine on Chrome.

**Comment by cnishina** [comment]
We recommend against using PhantomJS for tests with Protractor. Please see the browser support page. http://www.protractortest.org/#/browser-support

Possibly a good place is to ask this questions on [StackOverflow](http://stackoverflow.com/questions/tagged/protractor), [Google Group discussion list](https://groups.google.com/forum/?fromgroups#!forum/angular), or [Gitter](https://gitter.im/angular/protractor).

Thank you!

**Comment by debanshuk** [comment]
@cnishina, I understand that protractor doesn't fully support PhantomJS. But the issue reported here has been marked as resolved in #85 with this comment https://github.com/angular/protractor/issues/85#issuecomment-33159865, stating that the fix would be out in version 0.17.0 (https://github.com/angular/protractor/commit/a0bd84b9a28ec92eccd2784f8b849388985a4480). This was in 2014. There was another commit fixing the issue again 2015, ie, https://github.com/angular/protractor/commit/d15ccdcf82bda29c803c3a5642896f16d7e4f938. Now somehow the issue has reappeared. I.e, this is a use case was supported by Protractor earlier, but now you are saying it isn't?

**Comment by debanshuk** [comment]
Can someone add label 'browser: phantomjs' on this and all the other PhantomJS issues stated in the description. So that people can find them with this link: https://github.com/mattermost/desktop/blob/master/package.json. This is the link present in know issues column of PhantomJS section in browser support page. It currently one have one of the issues present in description and doesn't contain the most important issue, i.e. #85.

Also, some kind of won't fix label should be added to these issues, as new people trying PhantomJS might thing that these closed issues have been resolved, even when they are not.
